api/python/quilt3/data_transfer.py

```diff
@@ -890,15 +951,123 @@ def get_size_and_version(src: PhysicalKey):
     return size, version
 
 
-def calculate_sha256(src_list: List[PhysicalKey], sizes: List[int]):
+def calculate_checksum(src_list: List[PhysicalKey], sizes: List[int]) -> List[bytes]:
     assert len(src_list) == len(sizes)
 
     if not src_list:
         return []
-    return _calculate_sha256_internal(src_list, sizes, [None] * len(src_list))
+    return _calculate_checksum_internal(src_list, sizes, [None] * len(src_list))
 
 
-def _calculate_hash_get_s3_chunks(ctx, src, size):
+def with_lock(f):
+    lock = threading.Lock()
+
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        with lock:
+            return f(*args, **kwargs)
+    return wrapper
+
+
+@retry(stop=stop_after_attempt(MAX_FIX_HASH_RETRIES),
+       wait=wait_exponential(multiplier=1, min=1, max=10),
+       retry=retry_if_result(lambda results: any(r is None or isinstance(r, Exception) for r in results)),
+       retry_error_callback=lambda retry_state: retry_state.outcome.result(),
+       )
+def _calculate_checksum_internal(src_list, sizes, results) -> List[bytes]:
+    total_size = sum(
+        size
+        for size, result in zip(sizes, results)
+        if result is None or isinstance(result, Exception)
+    )
+    stopped = False
+
+    with tqdm(desc="Hashing", total=total_size, unit='B', unit_scale=True, disable=DISABLE_TQDM) as progress, \
+         ThreadPoolExecutor(MAX_CONCURRENCY) as executor:
+
+        find_correct_client = with_lock(S3ClientProvider().find_correct_client)
+        progress_update = with_lock(progress.update)
+
+        def _process_url_part(src: PhysicalKey, offset: int, length: int):
+            hash_obj = hashlib.sha256()
+
+            if src.is_local():
+                bytes_remaining = length
+                with open(src.path, 'rb') as fd:
+                    fd.seek(offset)
+                    while bytes_remaining > 0:
+                        chunk = fd.read(min(s3_transfer_config.io_chunksize, bytes_remaining))
+                        if not chunk:
+                            # Should not happen, but let's not get stuck in an infinite loop.
+                            break
```

Comment:
Well, this won't catch the case where file has changed, but became longer. We also have the same problem for S3 files, but it's less obvious. In any case, we'll need to come up with a "real" solution.

Reply:
Another solution is to say that we don't care what the user does to the file, but whatever ends up in S3 is what we consider to be "correct", and use its hash in the manifest.

In that case, we should only use `bytes_remaining` for the progress bar, but keep reading the file to the end.

Reply:
I did some more digging to figure out the correct solution.

Because we now rely on server-side checksums, this code only runs for `Package.verify` (in which case, it'll fail later anyway, so that's fine), and when the user explicitly runs `Package.build` (and I'm not sure when that's actually useful?)

If the user _does_ run `build` before `push`, then we should also worry about files changing _between_ `build` and `push`, not just during `build`. This can result all kinds of problems:
- `push` fails with a weird exception because the number of chunks decreased, and we tried doing a 0-byte part upload
- `push` succeeds, and returns a package with correct checksums but incorrect size (if the file got smaller)
- `push` succeeds, and returns a package with truncated files (if the file got larger)

We _could_ detect the case of files changing between `build` and `push` (rather than during) and handle that more gracefully. But we should decide what the expected behavior is, or even what the purpose of `build` is. E.g., suppose the file contents changes but the size stays the same; we detect that, but only _after_ the push is done - now what?

Reply:
(Anyways, I changed it to raise an exception.)

Reply:
> Because we now rely on server-side checksums, this code only runs for Package.verify (in which case, it'll fail later anyway, so that's fine), and when the user explicitly runs Package.build (and I'm not sure when that's actually useful?)

I'd also expect us to hash local file when ETag check succeeds and remote file doesn't have ChecksumSHA256, but IIUC now we end up hashing from S3. If my understanding is correct that looks like regression and it should be fixed.

Reply:
> I'd also expect us to hash local file when ETag check succeeds and remote file doesn't have ChecksumSHA256, but IIUC now we end up hashing from S3. If my understanding is correct that looks like regression and it should be fixed.

i'd say this is important enough to be fixed soon (in a follow-up (pre?)release), but not important enough to block the current PR and (pre)release

Reply:
Hmm interesting. I guess we could calculate the local SHA256 right there, and also use it to check SHA256 before we check the ETag, as @sir-sigurd suggested elsewhere. But yes, that's a big enough change that I'd rather do separately.